src/cpp/server/server_context.cc

```diff
@@ -54,12 +53,154 @@
+      }
+      return *this;
+    }
+    mem_utilization_.store(value, std::memory_order_relaxed);
+    if (GRPC_TRACE_FLAG_ENABLED(grpc_backend_metric_state_trace)) {
+      gpr_log(GPR_INFO, "[%p] Mem utilization recorded: %f", this, value);
+    }
+    return *this;
+  }
+  CallMetricRecorder& RecordQpsMetric(double value) override {
+    if (!IsQpsValid(value)) {
+      if (GRPC_TRACE_FLAG_ENABLED(grpc_backend_metric_state_trace)) {
+        gpr_log(GPR_INFO, "[%p] QPS value rejected: %f", this, value);
+      }
+      return *this;
+    }
+    qps_.store(value, std::memory_order_relaxed);
+    if (GRPC_TRACE_FLAG_ENABLED(grpc_backend_metric_state_trace)) {
+      gpr_log(GPR_INFO, "[%p] QPS recorded: %f", this, value);
+    }
+    return *this;
+  }
+  CallMetricRecorder& RecordUtilizationMetric(string_ref name,
+                                              double value) override {
+    if (!IsUtilizationValid(value)) {
+      if (GRPC_TRACE_FLAG_ENABLED(grpc_backend_metric_state_trace)) {
+        gpr_log(GPR_INFO, "[%p] Utilization value rejected: %s %f", this,
+                std::string(name.data(), name.length()).c_str(), value);
+      }
+      return *this;
+    }
+    internal::MutexLock lock(&mu_);
+    absl::string_view name_sv(name.data(), name.length());
+    utilization_[name_sv] = value;
+    if (GRPC_TRACE_FLAG_ENABLED(grpc_backend_metric_state_trace)) {
+      gpr_log(GPR_INFO, "[%p] Utilization recorded: %s %f", this,
+              std::string(name_sv).c_str(), value);
+    }
+    return *this;
+  }
+  CallMetricRecorder& RecordRequestCostMetric(string_ref name,
+                                              double value) override {
+    internal::MutexLock lock(&mu_);
+    absl::string_view name_sv(name.data(), name.length());
+    request_cost_[name_sv] = value;
+    if (GRPC_TRACE_FLAG_ENABLED(grpc_backend_metric_state_trace)) {
+      gpr_log(GPR_INFO, "[%p] Request cost recorded: %s %f", this,
+              std::string(name_sv).c_str(), value);
+    }
+    return *this;
+  }
+
+ private:
+  void GetBackendMetricData(grpc_core::BackendMetricData* data) override {
+    // Merge metrics from the ServerMetricRecorder first since metrics recorded
+    // to this (CallMetricRecorder) takes a higher precedence.
+    if (server_metric_recorder_) {
+      server_metric_recorder_->GetMetrics(data);
+    }
+    // Only overwrite if the value is set i.e. in the valid range.
+    const double cpu = cpu_utilization_.load(std::memory_order_relaxed);
+    if (IsUtilizationValid(cpu)) {
+      data->cpu_utilization = cpu;
+    }
+    const double mem = mem_utilization_.load(std::memory_order_relaxed);
+    if (IsUtilizationValid(mem)) {
+      data->mem_utilization = mem;
+    }
+    const double qps = qps_.load(std::memory_order_relaxed);
+    if (IsQpsValid(qps)) {
+      data->qps = qps;
+    }
+    {
+      internal::MutexLock lock(&mu_);
+      if (!utilization_.empty() || !request_cost_.empty()) {
```

Comment:
This check doesn't seem useful.  If the maps are empty, then the loops below will simply exit immediately.

Reply:
I intended to put the lock inside the if, updated and PTAL. The rational is that since we don't guarantee strict synchronization between set/get, we can return without acquiring the lock if both are empty. E.g. in prod I expect we'll return here in a big majority of cases.

Reply:
I changed to swap and left a comment saying the getter clears metrics.

And I think we can have both, let me explain if it makes sense. Map's const methods are safe to call without a lock so calling empty() itself should be safe. Also since we already do not guarantee synchronization between fields, checking whether maps are empty one at a time is no different than loading atomic fields one at a time to me.

Reply:
> Map's const methods are safe to call without a lock

As far as I know, that is an incorrect statement.  I certainly do not see anything in the API definition that says that this would be safe.

I think the only way this could possibly be true is if the implementation of `std::map<>` internally used an atomic to store the size of the map, because otherwise it's possible for one thread to be in the middle of adding an element while another thread uses a const pointer to call `empty()`.  And I very much doubt that any implementation would use an atomic for that, since that would impose overhead on every caller, even those that don't actually need the thread-safety that this would provide.

Reply:
Right, it's safe only among const methods, updated.